File: quicksort/cpp/multi_thread.cpp
#include <iostream>
#include <memory>
#include <array>
#include <random>
#include <stack>
#include <thread>

#include "hourmeter.h"

using namespace std;

#define N 20000000
#define RANGE 1000000000
#define NUM_THREADS 16
#define REPEAT 5

void quickSort(int arr[N], const int& start, const int& end) {
    int pivot, tmp;
    int i, j;
    int low, high;

    stack<int> stk;
    
    low = start;
    high = end;
    stk.push(high);
    stk.push(low);
    while (!stk.empty()) {
        low = stk.top();
        stk.pop();
        high = stk.top();
        stk.pop();

        if (low >= high) {
            continue;
        }
        int curLength = high - low + 1;
        int pivot;

        if (curLength > 2) {
            int tmpMid = (low + high) / 2;
            pivot = arr[low] > arr[high] ? (arr[low] > arr[tmpMid] ? (arr[tmpMid] > arr[high] ? tmpMid : high) : low) : (arr[high] > arr[tmpMid] ? (arr[tmpMid] > arr[low] ? tmpMid : low) : high);
            {
                tmp = arr[high];
                arr[high] = arr[pivot];
                arr[pivot] = tmp;
            }
            pivot = high;
        }
        else {
            if (arr[low] > arr[high]) {
                tmp = arr[low];
                arr[low] = arr[high];
                arr[high] = tmp;
            }
            continue;
        }

        //INFO: 데이터가 최소 두개
        pivot = arr[high];
        i = low - 1;
        j = high;
        while (1) {
            while (arr[++i] < pivot);
            while (arr[--j] > pivot);
            if (i >= j)
                break;
            tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
        }

        tmp = arr[i];
        arr[i] = arr[high];
        arr[high] = tmp;

        stk.push(high);
        stk.push(i + 1);
        stk.push(i - 1);
        stk.push(low);
    }
}

void main() {

	HourMeter hm;
	//INFO: 시드값을 얻기 위한 random_device 생성
	random_device rd;
	//INFO: random_device를 통해 난수 생성엔진을 초기화한다.
	mt19937 gen(rd());
	//INFO: 균등 분포 정의
	uniform_int_distribution<int> dis(0, RANGE);

    double timeSum = 0.0;

    for (int j = 0; j < REPEAT; j++) {
        int* arr = new int[N];

        for (int i = 0; i < N; i++) {
            arr[i] = dis(gen);
        }

        int threadJobSize = N / NUM_THREADS;
        printf("threadJobSize : %d\n", threadJobSize);

        thread thr[NUM_THREADS];

        hm.startMeasure();
        for (int i = 0; i < NUM_THREADS; i++) {
            int startPos = threadJobSize * i;
            int endPos = startPos + threadJobSize - 1;
            thr[i] = thread(quickSort, arr, startPos, endPos);
        }

        for (int i = 0; i < NUM_THREADS; i++) {
            if (thr[i].joinable()) {
                thr[i].join();
            }
        }

        hm.endMeasure();
        timeSum += hm.getLatestDuration();

        /*for (int i = 0; i < N; i++) {
            printf("%d ", arr[i]);
        }
        printf("\n");*/
        //printf("time : %lf\n", hm.getLatestDuration());

        delete[] arr;
    }

    printf("average : %lf\n", timeSum / REPEAT);
}

File: quicksort/go/multi_thread.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"sync"
	"time"
)

type Stack struct {
	items []int
}

func (s *Stack) Push(item int) {
	s.items = append(s.items, item)
}

func (s *Stack) Pop() int {
	dLen := len(s.items)
	if dLen == 0 {
		return -1
	}
	item, items := s.items[dLen-1], s.items[0:dLen-1]
	s.items = items
	return item
}

func (s *Stack) Empty() bool {
	dLen := len(s.items)
	if dLen == 0 {
		return true
	} else {
		return false
	}
}

func quickSort(wg *sync.WaitGroup, arr []int, startPos int, endPos int) {
	defer wg.Done()

	stk := Stack{}

	low := startPos
	high := endPos
	var pivot int
	var tmp int
	var i int
	var j int
	stk.Push(high)
	stk.Push(low)
	for stk.Empty() == false {
		low = stk.Pop()
		high = stk.Pop()
		curLength := high - low + 1
		if low >= high {
			continue
		}
		if curLength > 2 {
			tmpMid := int(math.Round(float64((low + high)) / 2))
			if arr[low] > arr[high] {
				if arr[low] > arr[tmpMid] {
					if arr[tmpMid] > arr[high] {
						pivot = tmpMid
					} else {
						pivot = high
					}
				} else {
					pivot = low
				}
			} else {
				if arr[high] > arr[tmpMid] {
					if arr[tmpMid] > arr[low] {
						pivot = tmpMid
					} else {
						pivot = low
					}
				} else {
					pivot = high
				}
			}
			tmp = arr[high]
			arr[high] = arr[pivot]
			arr[pivot] = tmp
			pivot = high
		} else {
			if arr[low] > arr[high] {
				tmp = arr[low]
				arr[low] = arr[high]
				arr[high] = tmp
			}
			continue
		}

		pivot = arr[high]
		i = low - 1
		j = high
		for true {
			for true {
				i++
				if arr[i] >= pivot {
					break
				}
			}
			for true {
				j--
				if arr[j] <= pivot {
					break
				}
			}
			if i >= j {
				break
			}
			tmp = arr[i]
			arr[i] = arr[j]
			arr[j] = tmp
		}
		tmp = arr[i]
		arr[i] = arr[high]
		arr[high] = tmp

		stk.Push(high)
		stk.Push(i + 1)
		stk.Push(i - 1)
		stk.Push(low)
	}
}

func main() {
	if len(os.Args) != 4 {
		panic("argv : JOB_SIZE MAX_VALUE NUM_THREADS")
	}

	//INFO: 데이터 개수
	jobSize, _ := strconv.Atoi(os.Args[1])
	//INFO: 데이터 최대 값
	maxValue, _ := strconv.Atoi(os.Args[2])
	//INFO: 스레드 개수
	numThreads, _ := strconv.Atoi(os.Args[3])
	var PrintCount int = 100

	//INFO: 값이 저장되는 배열
	arr := make([]int, jobSize)

	for i := 0; i < jobSize; i++ {
		arr[i] = rand.Intn(maxValue)
	}

	println("jobSize :", jobSize, ", maxValue :", maxValue, ", numThreads :", numThreads)

	var newWg sync.WaitGroup
	newWg.Add(numThreads)

	threadJobSize := jobSize / numThreads

	println("quicksort START")
	startTime := time.Now()
	for i := 0; i < numThreads; i++ {
		startPos := threadJobSize * i
		endPos := startPos + threadJobSize - 1
		//println("i: ", i, ", startPos: ", startPos, ", endPos: ", endPos)
		go quickSort(&newWg, arr, startPos, endPos)
	}
	newWg.Wait()
	elapsedTime := time.Since(startTime)
	println("quicksort END")
	d2 := elapsedTime / time.Millisecond
	et := float64(d2) / 1000.0

	fmt.Printf("elapsed time : %f sec\n", et)

	term := jobSize / PrintCount
	for i := 0; i < jobSize; i += term {
		println(arr[i])
	}

	filePath := "./result_go.txt"
	writeFile, err := os.OpenFile(filePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		panic(filePath + " file open failed")
	}
	defer writeFile.Close()
	var elapsedTimeStr string
	elapsedTimeStr = fmt.Sprintf("%f\n", et)
	if _, err := writeFile.WriteString(elapsedTimeStr); err != nil {
		panic(filePath + " file write failed")
	}
}
